File: trunk/src/org/simbrain/world/textworld/TextWorldDesktopComponent.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.world.textworld;

import java.awt.Dimension;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

import javax.swing.JMenu;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.KeyStroke;

import org.simbrain.util.SFileChooser;
import org.simbrain.workspace.gui.CouplingMenuComponent;
import org.simbrain.workspace.gui.GenericFrame;
import org.simbrain.workspace.gui.GuiComponent;

/**
 * <b>TextWorldComponent</b> is the container for the world component. Handles
 * toolbar buttons, and serializing of world data. The main environment codes is
 * in {@link TextWorld}.
 */
public class TextWorldDesktopComponent extends GuiComponent<TextWorldComponent> {

    /** Script directory. */
    private static final String SCRIPT_MENU_DIRECTORY = "."
            + System.getProperty("file.separator") + "scripts"
            + System.getProperty("file.separator") + "scriptmenu";

    private static final long serialVersionUID = 1L;

    /** Default height. */
    private static final int DEFAULT_HEIGHT = 450;

    /** Default width. */
    private static final int DEFAULT_WIDTH = 500;

    /** Menu Bar. */
    private JMenuBar menuBar = new JMenuBar();

    /** File menu for saving and opening world files. */
    private JMenu file = new JMenu("File  ");

    /** Opens an existing world file. */
    private JMenuItem open = new JMenuItem("Open");

    /** Saves the world. */
    private JMenuItem save = new JMenuItem("Save");

    /** Saves the world as a new file name. */
    private JMenuItem saveAs = new JMenuItem("Save As");

    /** Closes the current world. */
    private JMenuItem close = new JMenuItem("Close");

    /** Edit menu Item. */
    private JMenu edit = new JMenu("Edit  ");

    /** Opens the dialog to define TextWorld Dictionary. */
    private JMenu dictionary = new JMenu("Dictionary");

    /** Opens the dialog to define TextWorld Dictionary. */
    private JMenuItem loadDictionary = new JMenuItem("Load dictionary");

    /** Opens user preferences dialog. */
    private JMenuItem preferences = new JMenuItem("Preferences");

    /** Opens the help dialog for TextWorld. */
    private JMenu help = new JMenu("Help");

    /** Current file. */
    private File scriptFile;

    /** The pane representing the text world. */
    private TextWorldPanel panel;

    /**
     * Creates a new frame of type TextWorld.
     */
    public TextWorldDesktopComponent(GenericFrame frame,
            TextWorldComponent component) {
        super(frame, component);

        panel = new TextWorldPanel(component.getWorld());
        panel.setPreferredSize(new Dimension(DEFAULT_WIDTH, DEFAULT_HEIGHT));
        this.setPreferredSize(new Dimension(DEFAULT_WIDTH, DEFAULT_HEIGHT));
        addMenuBar();
        add(panel);
        frame.pack();
    }

    @Override
    public void postAddInit() {
        super.postAddInit();
        this.getParentFrame().pack();
    }

    /**
     * Adds menu bar to the top of TextWorldComponent.
     */
    private void addMenuBar() {

        // open.setActionCommand("open");
        open.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent arg0) {
                SFileChooser fileChooser = new SFileChooser(
                        SCRIPT_MENU_DIRECTORY, "Open file");
                scriptFile = fileChooser.showOpenDialog();
                TextWorldDesktopComponent.this.getParentFrame().setTitle(
                        scriptFile.getName());
                StringBuffer fileData = new StringBuffer(1000);
                BufferedReader reader;
                try {
                    reader = new BufferedReader(new FileReader(scriptFile));
                    char[] buf = new char[1024];
                    int numRead = 0;
                    while ((numRead = reader.read(buf)) != -1) {
                        String readData = String.valueOf(buf, 0, numRead);
                        fileData.append(readData);
                        buf = new char[1024];
                    }
                    reader.close();
                    // textArea.setText(fileData.toString());
                    panel.setInputText(fileData.toString());
                } catch (FileNotFoundException e) {
                    // TODO Auto-generated catch block
                    e.printStackTrace();
                } catch (IOException e) {
                    // TODO Auto-generated catch block
                    e.printStackTrace();
                }
            }
        });
        open.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_O, Toolkit
                .getDefaultToolkit().getMenuShortcutKeyMask()));
        save.setActionCommand("save");
        save.addActionListener(new ActionListener() {

            public void actionPerformed(ActionEvent e) {
                // TODO Auto-generated method stub
                try {
                    BufferedWriter out = new BufferedWriter(new FileWriter(
                            scriptFile));
                    // out.write(textArea.getText());
                    out.close();
                } catch (IOException exception) {
                    System.out.println("Exception ");

                }
            }
        });
        save.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_S, Toolkit
                .getDefaultToolkit().getMenuShortcutKeyMask()));
        close.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_W, Toolkit
                .getDefaultToolkit().getMenuShortcutKeyMask()));
        menuBar.add(file);
        file.add(open);
        file.add(save);
        file.add(saveAs);
        file.add(close);

        loadDictionary.setActionCommand("loadDictionary");
        loadDictionary.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_L,
                Toolkit.getDefaultToolkit().getMenuShortcutKeyMask()));
        // preferences.setActionCommand("prefs");
        preferences.addActionListener(new ActionListener() {

            public void actionPerformed(ActionEvent e) {
                panel.showTextWorldDialog();
            }
        });

        preferences.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_P,
                Toolkit.getDefaultToolkit().getMenuShortcutKeyMask()));
        menuBar.add(edit);
        dictionary.add(loadDictionary);
        edit.add(dictionary);
        edit.add(preferences);

        menuBar.add(new CouplingMenuComponent("Couple", this.getWorkspaceComponent()
                .getWorkspace(), this.getWorkspaceComponent()));

        menuBar.add(help);

        getParentFrame().setJMenuBar(menuBar);
    }


    @Override
    public void closing() {
        // TODO Auto-generated method stub

    }

}

File: src/main/kotlin/org/simbrain/world/textworld/gui/TextWorldDesktopComponent.kt
package org.simbrain.world.textworld.gui

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.swing.Swing
import org.simbrain.util.genericframe.GenericFrame
import org.simbrain.util.widgets.ShowHelpAction
import org.simbrain.workspace.couplings.getProducer
import org.simbrain.workspace.gui.CouplingMenu
import org.simbrain.workspace.gui.DesktopComponent
import org.simbrain.workspace.gui.SimbrainDesktop
import org.simbrain.world.textworld.*
import java.awt.Dimension
import java.awt.event.ComponentAdapter
import java.awt.event.ComponentEvent
import javax.swing.JMenu
import javax.swing.JMenuBar
import javax.swing.JMenuItem

/**
 * **ReaderComponentDesktopGui** is the gui view for the reader world.
 */
class TextWorldDesktopComponent(frame: GenericFrame, component: TextWorldComponent) :
    DesktopComponent<TextWorldComponent>(frame, component) {
    /**
     * Menu Bar.
     */
    private val menuBar = JMenuBar()

    /**
     * File menu for saving and opening world files.
     */
    private val file = JMenu("File")

    /**
     * Edit menu Item.
     */
    private val edit = JMenu("Edit")

    /**
     * Opens user preferences dialog.
     */
    private val preferences = JMenuItem("Preferences")

    /**
     * Opens the help dialog for TextWorld.
     */
    private val help = JMenu("Help")

    /**
     * Help menu item.
     */
    private val helpItem = JMenuItem("Reader Help")

    /**
     * The pane representing the text world.
     */
    val panel: TextWorldPanel

    /**
     * The text world.
     */
    private val world: TextWorld

    /**
     * Creates a new frame of type TextWorld.
     *
     * @param frame
     * @param component
     */
    init {
        world = component.world
        panel = TextWorldPanel(world)
        this.preferredSize = Dimension(DEFAULT_WIDTH, DEFAULT_HEIGHT)
        addMenuBar()
        add(panel)
        frame.pack()

        val updater = component.workspace.updater
        updater.events.runStarted.on(Dispatchers.Swing) { panel.setRunLock(world.lockWhileRunning) }
        updater.events.runFinished.on(Dispatchers.Swing) { panel.setRunLock(false) }
        panel.setRunLock(world.lockWhileRunning && updater.isRunning)

        // Force component to fill up parent panel
        addComponentListener(object : ComponentAdapter() {
            override fun componentResized(e: ComponentEvent) {
                val component = e.component
                panel.preferredSize = Dimension(component.width, component.height)
                panel.revalidate()
            }
        })
        parentFrame.pack()
    }

    /**
     * Adds menu bar to the top of TextWorldComponent.
     */
    private fun addMenuBar() {

        // File Menu
        menuBar.add(file)
        file.add(SimbrainDesktop.actionManager.createImportAction(this))
        file.add(SimbrainDesktop.actionManager.createExportAction(this))
        file.addSeparator()
        file.add(loadTextAction)
        file.addSeparator()
        file.add(SimbrainDesktop.actionManager.createRenameAction(this))
        file.addSeparator()
        file.add(SimbrainDesktop.actionManager.createCloseAction(this))

        // Edit menu
        preferences.action = world.textWorldPrefs
        edit.add(createShowFindAndReplaceAction())
        edit.addSeparator()
        edit.add(
            SimbrainDesktop.actionManager.createCoupledDataWorldAction(
                name = "Record token embeddings",
                world.getProducer(TextWorld::currentVector),
                sourceName = "${world.id} Token Embeddings",
                world.tokenEmbedding.dimension
            )
        )
        edit.add(
            SimbrainDesktop.actionManager.createCoupledPlotMenu(
                world.getProducer(TextWorld::currentVector),
                "${world.id} Token Embeddings",
            )
        )
        edit.addSeparator()
        edit.add(CouplingMenu(workspaceComponent, world))
        edit.addSeparator()
        edit.add(preferences)
        menuBar.add(edit)

        // View Menu
        val viewMenu = JMenu("View")
        viewMenu.add(world.extractEmbeddingFromCurrentText)
        viewMenu.add(world.viewTokenEmbedding)
        menuBar.add(viewMenu)

        // Help Menu
        menuBar.add(help)
        val helpAction = ShowHelpAction("https://docs.simbrain.net/docs/worlds/textworld.html")
        helpItem.action = helpAction
        help.add(helpItem)

        // Add menu
        parentFrame.jMenuBar = menuBar
    }

    companion object {
        /**
         * Default height.
         */
        private const val DEFAULT_HEIGHT = 250

        /**
         * Default width.
         */
        private const val DEFAULT_WIDTH = 400
    }
}